PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/test/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/HardcodedAPIKeysAndTokensCheckTest.java

```diff
@@ -0,0 +1,134 @@
+     */
+    @Test
+    public void testHardcodedAPIKeysAndTokensCheck() {
+
+        // assert visitor
+        assertVisitor(mockVisitor);
+
+        int NUMBER_OF_INVOCATIONS = 1; // number of times registerProblem is called
+
+        String CLIENT_TO_CHECK_ONE = "AccessToken";
+        verifyRegisterProblem(CLIENT_TO_CHECK_ONE, NUMBER_OF_INVOCATIONS);
+
+        String CLIENT_TO_CHECK_TWO = "AzureKeyCredential";
+        verifyRegisterProblem(CLIENT_TO_CHECK_TWO, NUMBER_OF_INVOCATIONS);
+    }
+
+    /**
+     * Test for non-auth client use.
+     * These are other Azure clients that are not AzureKeyCredential or AccessToken.
+     */
+    @Test
+    public void testNonAuthAzureClientUse() {
+
+        // assert visitor
+        assertVisitor(mockVisitor);
+
+        String CLIENT_TO_CHECK = "SomeOtherClient";
+        int NUMBER_OF_INVOCATIONS = 0;
+        verifyRegisterProblem(CLIENT_TO_CHECK, NUMBER_OF_INVOCATIONS);
+    }
+
+    /**
+     * Test for null class reference.
+     */
+    @Test
+    public void testNullClassReference() {
+
+        // assert visitor
+        assertVisitor(mockVisitor);
+
+        String CLIENT_TO_CHECK = ""; // null class reference
+        int NUMBER_OF_INVOCATIONS = 0;
+        verifyRegisterProblem(CLIENT_TO_CHECK, NUMBER_OF_INVOCATIONS);
+    }
+
+
+    // Helper method to create visitor.
+    PsiElementVisitor createVisitor() {
+        boolean isOnTheFly = true;
+        APIKeysAndTokensVisitor visitor = new HardcodedAPIKeysAndTokensCheck.APIKeysAndTokensVisitor(mockHolder, isOnTheFly);
+        return visitor;
+    }
+
+
+    // Helper method to assert visitor is not null and is an instance of JavaElementVisitor
+    void assertVisitor(PsiElementVisitor mockVisitor) {
+
+        assertNotNull(mockVisitor);
+        assertTrue(mockVisitor instanceof JavaElementVisitor);
+    }
+
+    // Helper method to verify registerProblem is called
+    void verifyRegisterProblem(String CLIENT_TO_CHECK, int NUMBER_OF_INVOCATIONS) {
+
+        PsiNewExpression newExpression = mock(PsiNewExpression.class);
+        PsiJavaCodeReferenceElement javaCodeReferenceElement = mock(PsiJavaCodeReferenceElement.class);
+
+        when(newExpression.getClassReference()).thenReturn(javaCodeReferenceElement);
+        when(javaCodeReferenceElement.getReferenceName()).thenReturn(CLIENT_TO_CHECK);
+
+        when(javaCodeReferenceElement.getQualifiedName()).thenReturn("com.azure");
+
+        mockVisitor.visitElement(newExpression);
+
+        // Verify registerProblem is called
+        verify(mockHolder, times(NUMBER_OF_INVOCATIONS)).registerProblem(eq(newExpression),
+                Mockito.contains("DefaultAzureCredential is recommended for authentication if the service client supports Token Credential (Entra ID Authentication). " +
+                        "If not, then use Azure Key Credential for API key based authentication."));
```

Comment:
This seems to be incorrect. We are flagging APIKey Credential pattern but also suggesting it?

Reply:
This is a reasonable message; we need to allow for circumstances where the user can't use one of the better options (services that only support keys, etc.)